### PR TITLE
Automatic Layouts and Rendering

### PR DESCRIPTION
**Pull request by dwaynepryce**
This is a new feature to provide some processes for automatic layout generation and rendering.

`graspologic/layouts/auto.py` is the main automatic do-a-layout file.  There are currently two dimensionality reduction processes we use; t-SNE and UMAP.  The process for each is:

- take a networkx graph
- shrink it if necessary by cutting the smallest edges
- use only the largest connected component
- create a node2vec embedding from it
- use tsne or umap to get an x and y coordinate from it
- automatically size each node based on their degree
- run a no-overlap process that @bryantower created which shuffles things around to make sure we don't have thousands of nodes on the same point
- create a color palette based on the leiden communities and a [thematic](https://microsoft.github.io/thematic) color palette and use it for the nodes in those communities
- ultimately using networkx' draw graph capabilities, save the graph 

We also added a runnable module so that you can call it via `python -m graspologic.layouts {args}` to run it from CLI instead of within the interpreter.  It's fairly on rails on purpose, because we wanted people to use the library not the CLI.

### CODE REVIEW

File: docs/reference/index.rst [hunk 1]
```diff
@@ -6,17 +6,19 @@ Reference
 .. toctree::
    :maxdepth: 2
 
-   simulations
-   models
+   align
```
Comment (dwaynepryce) [inline docs/reference/index.rst:9]: I couldn't really make sense of the ordering in this toctree so I made it alphabetical.  If we want to go back to the way it was that is fine, I just thought it might have been a legacy thing with no real good reason

**Review by bdpedigo — COMMENTED** [review]
I left a few small comments. A few bigger picture ones (which may be more for conversations down the road, but still):
- at some point, it would be nice to unify the interface for `.layouts` with `.plot`. Fundamentally I don't see a reason why they need to be different. I think this would just mean making the plotting functions conform more to a seaborn-style axes-level plotting function, and then still having wrappers to plot/save as needed when running from the command line. I think this would make it easier for newcomers to pick up the tools here. 
- I'll just say for now that I really don't know what's going on in the QuadTree/QuadNode/NoOverlap stuff, and don't have a publication or anything to look at, some kind of high level explanation for how this works would be helpful for me at least. 
   - {Side node:} Regarding the questions about this and the policy @jovo proposed, I brought it up with him and I think we decided we should just have a policy for which things can be included in the package that is fair, and should discuss further what that policy should be. Sounded like something we should discuss more in the new year.
- some example graphs + graph layouts would be pretty helpful to understand the utility.
- I suspect that if I were using this stuff, more access to the params of any of the underlying embedding functions would be helpful since they tend to make a huge difference in results.
- I'm a bit hesitant about the idea of the CLI since it is just not something most data science packages have. I realize it is helpful to the workflow at MSR and I don't want to hold this PR up generally but perhaps something we should also revisit going forward.
- I think worth noting in the docs that none of the renders will be deterministic as there are like 5 layers of randomness/non-identifiability (which is fine, but important to know as the user)

File: docs/reference/index.rst (hunk 1 above)
Comment (bdpedigo) [reply docs/reference/index.rst:9]: you are correct and I have no issue with this

File: graspologic/layouts/auto.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        for Dimension Reduction, ArXiv e-prints 1802.03426, 2018
+    .. [2] Böhm, Jan Niklas; Berens, Philipp; Kobak, Dmitry. A Unifying Perspective
+        on Neighbor Embeddings along the Attraction-Repulsion Spectrum. ArXiv e-prints
+        2007.08902v1, 17 Jul 2020.
+    """
+
+    lcc_graph, tensors, labels = _node2vec_for_layout(graph, max_edges)
+    points = umap.UMAP(min_dist=min_dist, n_neighbors=n_neighbors).fit_transform(
+        tensors
+    )
```
Comment (bdpedigo) [inline graspologic/layouts/auto.py:167]: I don't see this dependency being added to the package anywhere am I missing it somewhere?

[the remark this answers is not in the document]
Comment (dwaynepryce) [reply graspologic/layouts/auto.py:167]: I'm not opposed to it in the future, but I'm a big believer in limiting your capabilities at first and then relaxing them as you go - that way you don't claim something is going to work but you've never actually tested it, and so don't know whether it works or not.

That's why we took a fairly limited approach at first. :)

File: graspologic/layouts/auto.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+    largest_component = max(nx.connected_components(graph), key=len)
+    return graph.subgraph(largest_component).copy()
+
+
+def _approximate_prune(graph: nx.Graph, max_edges_to_keep: int = 1000000):
+    num_edges = len(graph.edges())
+    logger.info(f"num edges: {num_edges}")
+
+    if num_edges > max_edges_to_keep:
+        histogram, bins = histogram_edge_weight(graph, bin_directive=100)
```
[the remark this answers is not in the document]
Comment (dwaynepryce) [reply graspologic/layouts/auto.py:182]: Precisely, and bin directive could absolutely be exposed - I just didn't to start with since I'm taking the "let's not make this even more complicated than it has to be" initial tack

Binary search wouldn't work because all you have in the histogram at each index is the count in that bin.  That's why we do this kinda janky approximate approach here - we're counting the population in each bucket, starting with the largest weights and moving down.  Eventually we'll find a bucket that would tip us over our threshold, and we prune the entire bucket and everything smaller than it. 

I thought about exposing this in `graph_cuts.py`, but I would want something to actually find me precisely `max_edges_to_keep` and make the cuts to get to that point, which would require us to identify the edges within the bin, sort them, and prune the smallest until we hit that threshold.  And that is totally reasonable, but there's already so much in this PR I wasn't too keen on trying to do it right now, and so we went with this brute force approximate approach to start with.

File: graspologic/layouts/auto.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+        logger.debug(f"after cut num edges: {len(graph.edges())}")
+
+    return graph
+
+
+def _node2vec_for_layout(
+    graph: nx.Graph,
+    max_edges: int = 10000000,
+) -> Tuple[nx.Graph, np.ndarray, np.ndarray]:
+    graph = _approximate_prune(graph, max_edges)
```
[the remark this answers is not in the document]
Comment (dwaynepryce) [reply graspologic/layouts/auto.py:203]: wat

File: graspologic/layouts/auto.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+
+
+def _get_bounds(points: np.ndarray) -> Tuple[float, float, float, float]:
+    min_x, min_y, max_x, max_y = math.inf, math.inf, -math.inf, -math.inf
+    for x, y in points:
+        min_x = min(x, min_x)
+        min_y = min(y, min_y)
+        max_x = max(x, max_x)
+        max_y = max(y, max_y)
+    return min_x, min_y, max_x, max_y
```
[the remark this answers is not in the document]
Comment (dwaynepryce) [reply graspologic/layouts/auto.py:278]: I think @bryantower was presuming a single loop is better than two loops (one for max, one for min).

File: graspologic/layouts/auto.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+
+
+def _center_points_on_origin(points: np.ndarray) -> np.ndarray:
+    min_x, min_y, max_x, max_y = _get_bounds(points)
+    x_range = max_x - min_x
+    y_range = max_y - min_y
+    center_x = x_range / 2 + min_x
+    center_y = y_range / 2 + min_y
+    moved_points = [(x - center_x, y - center_y) for x, y in points]
```
[the remark this answers is not in the document]
Comment (dwaynepryce) [reply graspologic/layouts/auto.py:287]: This is a good catch - I tried to put most of these in np-isms but missed this one.  Thanks @bdpedigo !

File: graspologic/layouts/colors.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    (
+        light,
+        dark,
+    ) = _CACHED_LIGHT, _CACHED_DARK if theme_path is None else _load_thematic_json(
+        theme_path
+    )
+    return light if light_background else dark
+
+
+def nominal_colors(
```
[the remark this answers is not in the document]
Comment (dwaynepryce) [reply graspologic/layouts/colors.py:47]: @natoverse why do you call these "nominal" instead of "categorical" or "qualitative".

@bdpedigo I used nominal simply because that's what the json file that gives us our palette calls it.  And that's all on Nathan.

File: graspologic/layouts/render.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    figure_height: float = 15.0,
+    light_background: bool = True,
+    vertex_shape: str = "o",
+    arrows: bool = False,
+    dpi: int = 500,
+):
+    """
+    Renders and displays a graph.
+
+    Attempts to display it via the platform-specific display library such as TkInter
```
[the remark this answers is not in the document]
Comment (dwaynepryce) [reply graspologic/layouts/render.py:175]: So *draw* literally means it opens up a window and shows it to you, at that moment.  Matplotlib usually (always?) uses TkInter to do this, but you need to install it on some systems to make it work - I'm pretty sure on Ubuntu you literally have to install tkinter as an `apt` thing not just a straight python thing, since it's actually displaying it as a GUI type window for you and lets you zoom and pan and save and whatnot.  It's different on MacOS and Windows, but I don't know the super specifics.

The main point is that you shouldn't use this function unless you have a platform-specific display library like tkinter installed.

File: graspologic/layouts/render.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+    Parameters
+    ----------
+    graph : nx.Graph
+        The graph to be displayed. If the networkx Graph contains only nodes, no
+        edges will be displayed.
+    positions : List[:class:`graspologic.layouts.NodePosition`]
+        The positionsfor every node in the graph.
+    node_colors : Dict[Any, str]
+        A mapping of node id to colors. Must contain an entry for every node in the
+        graph.
```
[the remark this answers is not in the document]
Comment (dwaynepryce) [reply graspologic/layouts/render.py:186]: Right now they're colored with the source node color.  At one point they were all black and we changed that to be source-nodey, but it's just as valid to say target node coloring too.  I think it just needs a little bit more noodling before I'm ready to say it'll be good for everyone, but it's also not bad as a starting point either

**Comment by dwaynepryce** [comment]
> * at some point, it would be nice to unify the interface for `.layouts` with `.plot`. Fundamentally I don't see a reason why they need to be different. I think this would just mean making the plotting functions conform more to a seaborn-style axes-level plotting function, and then still having wrappers to plot/save as needed when running from the command line. I think this would make it easier for newcomers to pick up the tools here.

Maybe - I have a bit of reticence solely because when I plot something, I already know I have a graph that is worth doing something to, and while doing technique `x` on it may not be useful, I still know I have a graph that I can work with.  For us, one of our very first steps when we're handed a bundle of data is to try to render it and see if it actually _is_ worth doing something to it or not.

As far as the seaborn style stuff goes - I'm willing to look into it for sure.  We kept matplotlib because initially we did not support `seaborn` in our `topologic` release to try to minimize our number of dependencies.  Since GrasPY already was using seaborn, we can definitely explore changing it up.

> * I'll just say for now that I really don't know what's going on in the QuadTree/QuadNode/NoOverlap stuff, and don't have a publication or anything to look at, some kind of high level explanation for how this works would be helpful for me at least.

You and me both. That's all @bryantower and I'm pretty sure he's going to write something up at some point, I just don't know when.

>   * {Side node:} Regarding the questions about this and the policy @jovo proposed, I brought it up with him and I think we decided we should just have a policy for which things can be included in the package that is fair, and should discuss further what that policy should be. Sounded like something we should discuss more in the new year.

Sounds good!

> * some example graphs + graph layouts would be pretty helpful to understand the utility.

I agree - a tutorial for it would make a lot of sense, actually.  I'll create a ticket for it.

> * I suspect that if I were using this stuff, more access to the params of any of the underlying embedding functions would be helpful since they tend to make a huge difference in results.

Agreed, but our primary goal is to NOT have users tweaking things - we're looking for a "pretty good" rendering each and every time, not a bespoke render every time.  There's an argument to be made that what we should have for our `layout_tsne` and `layout_umap` functions is a "given this graph and these x, y coordinates, go do all the other coloring/sizing/grouping/no-overlapping" bit so you can pass in whatever embedding you want and run your own dimensionality reduction as you see fit prior to that. 

> * I'm a bit hesitant about the idea of the CLI since it is just not something most data science packages have. I realize it is helpful to the workflow at MSR and I don't want to hold this PR up generally but perhaps something we should also revisit going forward.

If it makes you feel any better, go into your venv and type `python -m json.tool -h` right now - that's from the default `json` module that comes with your python.  It may not be super common but it's still a pythonic thing and I don't think we should shy away from it just because other dsci packages don't make use of it :)

> * I think worth noting in the docs that none of the renders will be deterministic as there are like 5 layers of randomness/non-identifiability (which is fine, but important to know as the user)

I think one of the things we should do right at the beginning of the near year is allow us to seed this so we can get deterministic outputs.  Good note!

**Comment by dwaynepryce** [comment]
Created #612, #613, and #614

File: graspologic/layouts/auto.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bdpedigo) [reply graspologic/layouts/auto.py:167]: sounds good, I see #612, resolving

File: graspologic/layouts/render.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bdpedigo) [reply graspologic/layouts/render.py:186]: mostly wondering for doc purposes, so maybe this should say something about edge colors also for now?

File: graspologic/layouts/colors.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (natoverse) [reply graspologic/layouts/colors.py:47]: We use nominal more or less interchangeably with categorical. Sometimes people lump in ordinal with categorical, so it seems a bit more precise to use "nominal" to refer to plain labeled data like this, but honestly I don't think it matters much. The underlying thematic library uses "nominal", but feel free to use categorical when you map it in graspologic if it aligns better with your ecosystem :).

File: graspologic/layouts/auto.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bryantower) [reply graspologic/layouts/auto.py:278]: I believe that it was a list and not a numpy array originally.  It should probably switch to use numpy.max

File: graspologic/layouts/__main__.py [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+    graph, positions = auto.layout_tsne(
+        graph, perplexity=30, n_iter=1000, max_edges=arguments.max_edges
+    )
+    _output(arguments, graph, positions)
+
+
+def _umap(arguments: argparse.Namespace):
+    valid_args(arguments)
+    graph = _graph_from_file(arguments.edge_list, arguments.skip_header)
+    graph, positions = auto.layout_umap(graph, max_edges=arguments.max_edges)
```
Comment (bryantower) [inline graspologic/layouts/__main__.py:71]: This seems to be missing the ability to remove the overlap from nodes.  I think we need to have that added back in there.

File: graspologic/layouts/__main__.py [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+    if arguments.location_file is not None:
+        _location(arguments.location_file, positions, color_map)
+
+
+def _tsne(arguments: argparse.Namespace):
+    valid_args(arguments)
+    graph = _graph_from_file(arguments.edge_list, arguments.skip_header)
+    graph, positions = auto.layout_tsne(
+        graph, perplexity=30, n_iter=1000, max_edges=arguments.max_edges
+    )
```
Comment (bryantower) [inline graspologic/layouts/__main__.py:64]: This seems to be missing the ability to remove the overlap from nodes. I think we need to have that added back in there.

File: graspologic/layouts/auto.py [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+    return lcc_graph, positions
+
+
+def _largest_connected_component(graph: nx.Graph) -> nx.Graph:
+    largest_component = max(nx.connected_components(graph), key=len)
+    return graph.subgraph(largest_component).copy()
+
+
+def _approximate_prune(graph: nx.Graph, max_edges_to_keep: int = 1000000):
+    num_edges = len(graph.edges())
```
Comment (Nyecarr) [inline graspologic/layouts/auto.py:178]: Is this faster than sorting the edges by weight descending and just taking the top `max_edges_to_keep` nodes?

Comment (dwaynepryce) [reply graspologic/layouts/auto.py:178]: O(n) vs. O(nlogn), right?

File: graspologic/layouts/__main__.py (hunk 10 above)
Comment (dwaynepryce) [reply graspologic/layouts/__main__.py:71]: oh jeez - the no overlap doesn't go here, it goes in auto, but it was still missing from that.  derp.

File: graspologic/layouts/auto.py [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+        NodePosition(
+            node_id=key,
+            x=scaled_points[index][0],
+            y=scaled_points[index][1],
+            size=sizes[key],
+            community=partitions[key],
+        )
+        for index, key in enumerate(labels)
+    ]
+    positions = remove_overlaps(positions)
```
Comment (dwaynepryce) [inline graspologic/layouts/auto.py:237]: @bryantower we are removing overlaps now

**Review by bryantower — APPROVED** [review]
It looks like a good first version.  There is definitely room for improvement, but we can work on that in separate pull requests.

File: graspologic/layouts/auto.py (hunk 12 above)
Comment (bryantower) [reply graspologic/layouts/auto.py:180]: Sorting might be faster, because I assume that the histogram call has to do sort first to gather the histogram stats.

Comment (dwaynepryce) [reply graspologic/layouts/auto.py:180]: Nah, just multiple O(n) passes. Presuming it's a scalar, first you get min and max, divvy it up according to your bin directive (creating `bin_directive` subranges within the full min to max range), then you go through again and put them in whichever bin they belong in.  

I think in some cases sorting would be faster. I'd like to run some tests to see at some point, but I think this is fine for now

Comment (dwaynepryce) [reply graspologic/layouts/auto.py:180]: I think the big thing to keep in mind is that the histogram we're using in our edge cuts is fixed width bins, that's why it's two linear passes vs. an `nlogn`

File: setup.py [hunk 14]
```diff
@@ -49,7 +49,7 @@
     package_data={"version": [os.path.join("graspologic", "version", "version.txt")]},
     install_requires=[
         "anytree>=2.8.0",
-        "gensim",
+        "gensim>=3.8.0,<=3.9.0", #methods signatures changed in the 4.0.0beta release
```
[the remark this answers is not in the document]
Comment (bryantower) [reply setup.py:52]: we should leave it as it is for now, and we can do a "try this if version is ___ else try ___"
